panic when including stddef.h; unknown node type: 'ConstantExpr'

Hi @elliotchance,

I know you've taken a pause from the project. So just wanted to say that I very much appreciate the time you've put in thus far! It's quite impressive what `c2go` is already capable of.

Consider this an issue for "the future", if you ever feel like getting back into coding on the project, or if someone else picks up the torch.

The following example code generates a panic when running `c2go transpile a.c`:

Contents of `a.c`:

```c
#include <stddef.h>

int f() {
	return 42;
}
```

(The important line is the `stddef.h`, I just included the function `f` to have some output to verify that "something happened" during the transpilation).


And the corresponding panic is as follows (on rev 4de3ec97b368135e371d3b0667f6345a9267f43c):

```
panic: unknown node type: 'ConstantExpr 0x562cf36b7190 <col:34, col:55> 'unsigned long''

goroutine 26 [running]:
github.com/elliotchance/c2go/ast.Parse(0xc00018e6b7, 0x3c, 0x6afb99, 0x5)
	/home/u/goget/src/github.com/elliotchance/c2go/ast/ast.go:272 +0x32e7
main.convertLinesToNodes(0xc0000d6630, 0xa, 0x14, 0x0, 0x0, 0x0)
	/home/u/goget/src/github.com/elliotchance/c2go/main.go:89 +0x1a6
main.convertLinesToNodesParallel.func1.1(0xc0000a2240, 0xc000096168, 0xc0000d6630, 0xa, 0x14, 0x0)
	/home/u/goget/src/github.com/elliotchance/c2go/main.go:113 +0x53
created by main.convertLinesToNodesParallel.func1
	/home/u/goget/src/github.com/elliotchance/c2go/main.go:111 +0x12a
```

This is on Clang 9.0:

```
u@x220 ~> clang -v
clang version 9.0.1 
Target: x86_64-pc-linux-gnu
```

The header file `stddef.h` (included from `a.c`) includes `__stddef_max_align_t.h` (which is part of Clang), the contents of which is as follows:

```c
/*===---- __stddef_max_align_t.h - Definition of max_align_t for modules ---===
 *
 * Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 * See https://llvm.org/LICENSE.txt for license information.
 * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 *
 *===-----------------------------------------------------------------------===
 */

#ifndef __CLANG_MAX_ALIGN_T_DEFINED
#define __CLANG_MAX_ALIGN_T_DEFINED

#if defined(_MSC_VER)
typedef double max_align_t;
#elif defined(__APPLE__)
typedef long double max_align_t;
#else
// Define 'max_align_t' to match the GCC definition.
typedef struct {
  long long __clang_max_align_nonce1
      __attribute__((__aligned__(__alignof__(long long))));
  long double __clang_max_align_nonce2
      __attribute__((__aligned__(__alignof__(long double))));
} max_align_t;
#endif

#endif
```

Wish you all the best and a happy spring to come!

Cheers,
Robin

--- a/ast/ast.go
+++ b/ast/ast.go
@@ -92,6 +92,8 @@ func Parse(fullline string) Node {
 		return parseConstAttr(line)
 	case "ConstantArrayType":
 		return parseConstantArrayType(line)
+	case "ConstantExpr":
+		return parseConstantExpr(line)
 	case "ContinueStmt":
 		return parseContinueStmt(line)
 	case "CompoundAssignOperator":

--- a/ast/constant_expr.go
+++ b/ast/constant_expr.go
@@ -1,0 +1,46 @@
+package ast
+
+// ConstantExpr is a constant expression
+type ConstantExpr struct {
+	Addr       Address
+	Pos        Position
+	Type       string
+	ChildNodes []Node
+}
+
+func parseConstantExpr(line string) *ConstantExpr {
+	groups := groupsFromRegex(
+		"<(?P<position>.*)> '(?P<type>.*?)'",
+		line,
+	)
+
+	return &ConstantExpr{
+		Addr:       ParseAddress(groups["address"]),
+		Pos:        NewPositionFromString(groups["position"]),
+		Type:       groups["type"],
+		ChildNodes: []Node{},
+	}
+}
+
+// AddChild adds a new child node. Child nodes can then be accessed with the
+// Children attribute.
+func (n *ConstantExpr) AddChild(node Node) {
+	n.ChildNodes = append(n.ChildNodes, node)
+}
+
+// Address returns the numeric address of the node. See the documentation for
+// the Address type for more information.
+func (n *ConstantExpr) Address() Address {
+	return n.Addr
+}
+
+// Children returns the child nodes. If this node does not have any children or
+// this node does not support children it will always return an empty slice.
+func (n *ConstantExpr) Children() []Node {
+	return n.ChildNodes
+}
+
+// Position returns the position in the original source code.
+func (n *ConstantExpr) Position() Position {
+	return n.Pos
+}

--- a/ast/constant_expr_test.go
+++ b/ast/constant_expr_test.go
@@ -1,0 +1,18 @@
+package ast
+
+import (
+	"testing"
+)
+
+func TestConstantExpr(t *testing.T) {
+	nodes := map[string]Node{
+		`0x5558ffde68f0 <col:34, col:55> 'unsigned long'`: &ConstantExpr{
+			Addr:       0x5558ffde68f0,
+			Pos:        NewPositionFromString("col:34, col:55"),
+			Type:       "unsigned long",
+			ChildNodes: []Node{},
+		},
+	}
+
+	runNodeTests(t, nodes)
+}

--- a/ast/position.go
+++ b/ast/position.go
@@ -263,6 +263,8 @@ func setPosition(node Node, position Position) {
 		n.Pos = position
 	case *ConstAttr:
 		n.Pos = position
+	case *ConstantExpr:
+		n.Pos = position
 	case *ContinueStmt:
 		n.Pos = position
 	case *CompoundAssignOperator:

--- a/transpiler/literals.go
+++ b/transpiler/literals.go
@@ -109,3 +109,12 @@ func transpileCompoundLiteralExpr(n *ast.CompoundLiteralExpr, p *program.Program
 	expr, t, _, _, err := transpileToExpr(n.Children()[0], p, false)
 	return expr, t, err
 }
+
+func transpileConstantExpr(n *ast.ConstantExpr, p *program.Program) (goast.Expr, string, error) {
+	children := n.Children()
+	expr, t, _, _, err := transpileToExpr(children[0], p, false)
+	if len(children) != 1 {
+		p.AddMessage(p.GenerateWarningMessage(fmt.Errorf("ConstantExpr has %d children, expected 1 child", len(children)), n))
+	}
+	return expr, t, err
+}

--- a/transpiler/literals.go
+++ b/transpiler/literals.go
@@ -116,5 +116,8 @@ func transpileConstantExpr(n *ast.ConstantExpr, p *program.Program) (goast.Expr,
 	if len(children) != 1 {
 		p.AddMessage(p.GenerateWarningMessage(fmt.Errorf("ConstantExpr has %d children, expected 1 child", len(children)), n))
 	}
+	if len(n.Type) > 0 {
+		t = n.Type
+	}
 	return expr, t, err
 }

--- a/transpiler/transpiler.go
+++ b/transpiler/transpiler.go
@@ -152,6 +152,9 @@ func transpileToExpr(node ast.Node, p *program.Program, exprIsStmt bool) (
 		exprType = "double"
 		err = nil
 
+	case *ast.ConstantExpr:
+		expr, exprType, err = transpileConstantExpr(n, p)
+
 	case *ast.PredefinedExpr:
 		expr, exprType, err = transpilePredefinedExpr(n, p)
 

